packages/foam-vscode/src/features/create-from-template.ts

```diff
@@ -63,42 +63,64 @@ function findFoamVariables(templateText: string): string[] {
   return uniqVariables;
 }
 
-function getFoamTitle() {
+function resolveFoamTitle() {
   return window.showInputBox({
     prompt: `Enter a title for the new note`,
     value: 'Title of my New Note',
     validateInput: value =>
       value.trim().length === 0 ? 'Please enter a title' : undefined,
   });
 }
-
-function resolveFoamVariable(variable: string) {
-  if (variable === 'FOAM_TITLE') {
-    return getFoamTitle();
-  } else {
-    return Promise.resolve(variable);
+class Resolver {
+  promises = new Map<string, Thenable<string>>();
+
+  resolve(name: string, givenValues: Map<string, string>): Thenable<string> {
+    if (givenValues.has(name)) {
+      this.promises.set(name, Promise.resolve(givenValues.get(name)));
+    } else if (!this.promises.has(name)) {
+      switch (name) {
+        case 'FOAM_TITLE':
+          this.promises.set(name, resolveFoamTitle());
+          break;
+        default:
+          this.promises.set(name, Promise.resolve(name));
+          break;
+      }
+    }
+    const result = this.promises.get(name);
+    return result;
   }
 }
 
-function resolveFoamVariables(variables: string[]) {
+export async function resolveFoamVariables(
+  variables: string[],
+  givenValues: Map<string, string>
+) {
+  const resolver = new Resolver();
   const promises = variables.map(async variable =>
-    Promise.resolve([variable, await resolveFoamVariable(variable)])
+    Promise.resolve([variable, await resolver.resolve(variable, givenValues)])
   );
-  return Promise.all(promises);
-}
 
-export async function substituteFoamVariables(templateText: string) {
-  const variables = findFoamVariables(templateText);
-  const results = await resolveFoamVariables(variables);
+  const results = await Promise.all(promises);
 
   const valueByName = new Map<string, string>();
-  results.forEach(result => {
-    valueByName.set(result[0], result[1]);
+  results.forEach(([variable, value]) => {
+    valueByName.set(variable, value);
   });
 
-  variables.forEach(variable => {
-    const regex = new RegExp(`\\\${${variable}}|\\$${variable}`, 'g');
-    templateText = templateText.replace(regex, valueByName.get(variable));
+  return valueByName;
+}
+
+export function substituteFoamVariables(
+  templateText: string,
+  givenValues: Map<string, string>
+) {
+  givenValues.forEach((value, variable) => {
+    const regex = new RegExp(
+      `\\\${${variable}}|\\$${variable}([^_a-zA-Z0-9]|$)`,
```

Comment:
minor: I think I understand what you are doing here, I wonder if there is a better way to encode it

I would at least leave a comment to improve comprehension 

cc @jmg-duarte as he is a regex lover (and I hate them ;) )

Reply:
Sure thing. I've simplified the character class I copied from the [snippet syntax](https://code.visualstudio.com/docs/editor/userdefinedsnippets#_snippet-syntax), since it exactly matched `\W`, and added a comment that hopefully helps explain it.